format/FlightSQL.proto

```diff
@@ -0,0 +1,337 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto3";
+
+option java_package = "org.apache.arrow.flight.sql.impl";
+package arrow.flight.protocol.sql;
+
+/*
+ * Wrapper for the result of a "GetSQLInfo" action.
+ */
+message ActionGetSQLInfoResult {
+  /*
+   * Values are modelled after ODBC's SQLGetInfo() function. This information is intended to provide
+   * Flight SQL clients with basic, SQL syntax and SQL functions related information.
+   * More information types can be added in future releases.
+   * E.g. more SQL syntax support types, scalar functions support, type conversion support etc.
+   *
+   * Initially, Flight SQL will support the following information types:
+   * - Server Information
+   * -
+   *
+   * 1. Server Information: Provides basic information about the Flight SQL Server.
+   *
+   * The name of the Flight SQL Server.
+   * FLIGHT_SQL_SERVER_NAME
```

Comment:
I assume these are intended to be keys in the maps below?  For some of them should they be promoted to first class concepts?

Per my comment above.  I think I would have a preference for moving as much as possible to sql queries.  so if we defined a special table like 

"__APACHE_FLIGHT_SQL_INFO__" with a schema 
{ 
   flight_sql_server_name: utf8
   flight_sql_server_version: utf8 
   flight_sql_server_arrow_version: utf8 (or maybe modeled as a struct)
   flight_sql_info: map<utf8, dense_union<string_value: string, int_value: int32, bigint_value: int64, int32_bitmask: int32 >>
}
where the int32 bitmask would also have a defined extension metadata for a custom type of bitmask.

The minimal table requirement is that it supports projection:
i.e. only `select * from __APACHE_FLIGHT_SQL_INFO__` or `select  FLIGHT_SQL_VERSION from __APACHE_FLIGHT_SQL_INFO__`

Reply:
I think using pseudo tables is an interesting concept, with some interesting pros/cons worth discussing to decide on an approach.

Using pseudo tables allows for:
- a simple proto definition
- more expressiveness in terms of what can be done with the table, since it is then available for general SQL
- allowing later additions without needing to modify any proto

Drawbacks are mostly around two cases: those where metadata is not retrieved via a query mechanism and/or where it's more efficient not to use a query mechanism, and where you need to layer an additional parser in front of the actual database parser to support the pseudo. If we add a partial parser implementation to make it easier, we are having to implement this in all the languages that Flight SQL supports.

If we use the existing proposed proto definitions, then I see the following:
- less expressive in terms of what can be done
- less flexibility in terms of adding new metadata calls
- much less effort in that we don't need to provide a partial parser implementation to help
- It's easy to go from a message to a SQL query if that's what's used for metadata, it's hard to go the other way around

I think it would be easier all around to use what's defined here (with tweaks) vs. going the SQL table route, even though the SQL table route is more elegant from a protocol perspective. The effort to provide the hand-up for those data sources which don't have any support for querying metadata would be large, and may be a hinderance to adoption.

Reply:
I'll chime in and agree that while the pseudo tables are more elegant, it would make it drastically harder to support 'wrapping' existing databases without having to either modify the database implementation or pull in a SQL parser. Also, it sidesteps any concern or potential incompatibility about what subset of SQL is exactly supported.

Reply:
I like the idea of standardizing on information_schema proposed elsewhere, it seems a lot of DBs already support this, IIUC it is part of the SQL standard.  For those that don't support, I agree it might be a high barrier.   I'm a little concerned on cargo culting on one API standard or another which seems one level removed from the data we are trying to get at.

Reply:
for this particular feature, I'm not sure if information schema would supports it let me take a look more closely at this API.

Reply:
OK looked again.  For this RPC I'm OK with not requiring a pseudo-table here but I still think we should be returning this data as an arrow record batch instead of creating a new data model in proto accommodate it.  IIUC I think this captures some of benefits on the pseudo table approach but should be much easier to implement.  Thoughts?

Reply:
I'm fine with doing the record batch, that makes sense. As long as we can request multiple pieces of metadata at a time, then making the return format consistent is great.

The `information_schema` idea I think needs additional thought, I checked again and it doesn't indicate support for a catalog only table, so it might not be the silver bullet. We'll likely end up proposing additional tables in addition to what's there.

Reply:
Given that, I'd agree with @lidavidm that we should probably go with the initially proposed approach vs. pseudo tables.

Reply:
Yes, sounds good.  separately if I can find some time (probably unlikely), I'd like to see how hard it would be to make pseudo-tables work with Calcite and maybe ZetaSQL for c++.

Reply:
If we're going to return data as an Arrow record batch (I agree this makes sense) we should probably make all this data fetched through DoGet or DoExchange so that we're not specifying a separate encoding of Arrow data into Flight.

Reply:
We should also specify the expected schema inline here.

Reply:
Updated to use the existing DoGet mechanism.